tests/kernel/mem_protect/userspace/src/kobject.c

```diff
@@ -0,0 +1,166 @@
+/* 
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <ztest.h>
+#include <kernel_structs.h>
+#include <string.h>
+#include <stdlib.h>
+
+#define KOBJECT_STACK_SIZE 256
+
+/* Kernel objects */
+K_THREAD_STACK_DEFINE(kobject_test_4_stack, KOBJECT_STACK_SIZE);
+K_THREAD_STACK_DEFINE(kobject_test_5_1_stack, KOBJECT_STACK_SIZE);
+K_THREAD_STACK_DEFINE(kobject_test_5_2_stack, KOBJECT_STACK_SIZE);
+
+K_SEM_DEFINE(kobject_sem, 0, 10);
+K_MUTEX_DEFINE(kobject_mutex);
+__kernel struct k_thread kobject_test_4_tid;
+__kernel struct k_thread kobject_test_5_1_tid, kobject_test_5_2_tid;
+struct k_sem *random_sem_type=NULL;
+extern bool valid_fault;
+
+
+/****************************************************************************/
+void kobject_user_tc1(void *p1, void *p2, void *p3)
+{
+	valid_fault= true;
+	k_sem_take(random_sem_type, K_FOREVER);
+
+
+}
+
+void kobject_test_1(void *p1, void *p2, void *p3)
+{
+
+	/* _k_object_init(random_sem_type); */
+	k_thread_access_grant(k_current_get(),
+			      &kobject_sem,
+			      &kobject_mutex,
+			      random_sem_type, NULL);
+	
+	k_thread_user_mode_enter(kobject_user_tc1, NULL, NULL, NULL);
+
+}
+/****************************************************************************/
+void kobject_user_tc2(void *p1, void *p2, void *p3)
+{
+	valid_fault= false;
+	k_sem_give(&kobject_sem);
+
+	/* should cause a fault */
+	valid_fault= true;
+	k_sem_take((struct k_sem *)&kobject_mutex, K_FOREVER);
+
+
+}
+
+void kobject_test_2(void *p1, void *p2, void *p3)
+{
+	k_thread_access_grant(k_current_get(),
+			      &kobject_sem,
+			      &kobject_mutex, NULL);
+		
+	k_thread_user_mode_enter(kobject_user_tc2, NULL, NULL, NULL);
+	
+}
+
+/****************************************************************************/
+void kobject_user_tc3(void *p1, void *p2, void *p3)
+{
+	/* should cause a fault */
+	valid_fault= true;
+	k_sem_give(&kobject_sem);
+}
+
+void kobject_test_3(void *p1, void *p2, void *p3)
+{		
+	k_thread_user_mode_enter(kobject_user_tc3, NULL, NULL, NULL);
+	
+}
+
+/****************************************************************************/
+void kobject_user_test4(void *p1, void *p2, void *p3)
+{
+	/* should cause a fault */
+	if((u32_t)p1 == 1){
+		valid_fault= false;
+	}
+	else {
+		valid_fault= true;
+	}
+	k_sem_give(&kobject_sem);
+}
+
+void kobject_test_4(void *p1, void *p2, void *p3)
+{
+	k_thread_access_grant(k_current_get(),
+			      &kobject_sem, NULL);
+	
+	k_thread_create(&kobject_test_4_tid,
+			kobject_test_4_stack,
+			KOBJECT_STACK_SIZE,
+			kobject_user_test4,
+			(void *)1, NULL, NULL,
+			0, K_INHERIT_PERMS|K_USER, K_NO_WAIT);
+
+	k_sleep(1);
```

Comment:
We shouldn't rely on timing; explicitly synchronize if needed.

Reply:
Used instead of k_yield().